File: socialview/src/com/hendraanggrian/appcompat/widget/SocialView.java
package com.hendraanggrian.appcompat.widget;

import android.content.res.ColorStateList;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.PatternsCompat;

import com.hendraanggrian.appcompat.internal.SocialViewHelper;

import java.util.List;
import java.util.regex.Pattern;

/**
 * Base interface of all social widgets, which usually derived from {@link android.widget.TextView}.
 * Out of the box, there are {@link SocialTextView}, {@link SocialEditText}, and {@code SocialAutoCompleteTextView}.
 * It can also be implemented in any {@link android.view.View}
 * using {@link SocialViewHelper}.
 */
public interface SocialView {

    /**
     * Returns regex that are responsible for finding <b>hashtags</b>.
     * By default, the pattern are {@code #(\w+)}.
     */
    @NonNull
    Pattern getHashtagPattern();

    /**
     * Returns regex that are responsible for finding <b>mentions</b>.
     * By default, the pattern are {@code @(\w+)}.
     */
    @NonNull
    Pattern getMentionPattern();

    /**
     * Returns regex that are responsible for finding <b>hyperlinks</b>.
     * By default, the pattern are {@link PatternsCompat#WEB_URL}.
     */
    @NonNull
    Pattern getHyperlinkPattern();

    /**
     * Modify regex that are responsible for finding <b>hashtags</b>.
     *
     * @param pattern custom regex. When null, default pattern will be used.
     */
    void setHashtagPattern(@Nullable Pattern pattern);

    /**
     * Modify regex that are responsible for finding <b>hashtags</b>.
     *
     * @param pattern custom regex. When null, default pattern will be used.
     */
    void setMentionPattern(@Nullable Pattern pattern);

    /**
     * Modify regex that are responsible for finding <b>hashtags</b>.
     *
     * @param pattern custom regex. When null, default pattern will be used.
     */
    void setHyperlinkPattern(@Nullable Pattern pattern);

    /**
     * Returns true if <b>hashtags</b> in this view are spanned.
     */
    boolean isHashtagEnabled();

    /**
     * Returns true if <b>mentions</b> in this view are spanned.
     */
    boolean isMentionEnabled();

    /**
     * Returns true if <b>hyperlinks</b> in this view are spanned.
     */
    boolean isHyperlinkEnabled();

    /**
     * Determine whether this view should span <b>hashtags</b>.
     *
     * @param enabled True when spanning should be enabled.
     */
    void setHashtagEnabled(boolean enabled);

    /**
     * Determine whether this view should span <b>mentions</b>.
     *
     * @param enabled True when spanning should be enabled.
     */
    void setMentionEnabled(boolean enabled);

    /**
     * Determine whether this view should span <b>hyperlinks</b>.
     *
     * @param enabled True when spanning should be enabled.
     */
    void setHyperlinkEnabled(boolean enabled);

    /**
     * Returns color instance of <b>hashtags</b>, default is color accent of current app theme.
     * Will still return corresponding color even when {@link #isHashtagEnabled()} is false.
     */
    @NonNull
    ColorStateList getHashtagColors();

    /**
     * Returns color instance of <b>mentions</b>, default is color accent of current app theme.
     * Will still return corresponding color even when {@link #isMentionEnabled()} ()} is false.
     */
    @NonNull
    ColorStateList getMentionColors();

    /**
     * Returns color instance of <b>hyperlinks</b>, default is color accent of current app theme.
     * Will still return corresponding color even when {@link #isHyperlinkEnabled()} ()} is false.
     */
    @NonNull
    ColorStateList getHyperlinkColors();

    /**
     * Sets <b>hashtags</b> color instance.
     *
     * @param colors Colors state list instance.
     */
    void setHashtagColors(@NonNull ColorStateList colors);

    /**
     * Sets <b>mentions</b> color instance.
     *
     * @param colors Colors state list instance.
     */
    void setMentionColors(@NonNull ColorStateList colors);

    /**
     * Sets <b>hyperlinks</b> color instance.
     *
     * @param colors Colors state list instance.
     */
    void setHyperlinkColors(@NonNull ColorStateList colors);

    /**
     * Returns color integer of <b>hashtags</b>.
     *
     * @see #getHashtagColors()
     */
    @ColorInt
    int getHashtagColor();

    /**
     * Returns color integer of <b>mentions</b>.
     *
     * @see #getMentionColors()
     */
    @ColorInt
    int getMentionColor();

    /**
     * Returns color integer of <b>hyperlinks</b>.
     *
     * @see #getHyperlinkColors()
     */
    @ColorInt
    int getHyperlinkColor();

    /**
     * Sets <b>hashtags</b> color integer.
     *
     * @param color Color integer.
     * @see #setHashtagColors(ColorStateList)
     */
    void setHashtagColor(@ColorInt int color);

    /**
     * Sets <b>mentions</b> color integer.
     *
     * @param color Color integer.
     * @see #setMentionColors(ColorStateList)
     */
    void setMentionColor(@ColorInt int color);

    /**
     * Sets <b>hyperlinks</b> color integer.
     *
     * @param color Color integer.
     * @see #setHyperlinkColors(ColorStateList)
     */
    void setHyperlinkColor(@ColorInt int color);

    /**
     * Registers a callback to be invoked when a <b>hashtag</b> is clicked.
     *
     * @param listener The callback that will run.
     */
    void setOnHashtagClickListener(@Nullable OnClickListener listener);

    /**
     * Registers a callback to be invoked when a <b>mention</b> is clicked.
     *
     * @param listener The callback that will run.
     */
    void setOnMentionClickListener(@Nullable OnClickListener listener);

    /**
     * Registers a callback to be invoked when a <b>hyperlink</b> is clicked.
     *
     * @param listener The callback that will run.
     */
    void setOnHyperlinkClickListener(@Nullable OnClickListener listener);

    /**
     * Registers a text watcher to be invoked when a <b>hashtag</b> is modified.
     *
     * @param listener The callback that will run.
     */
    void setHashtagTextChangedListener(@Nullable OnChangedListener listener);

    /**
     * Registers a text watcher to be invoked when a <b>mention</b> is modified.
     *
     * @param listener The callback that will run.
     */
    void setMentionTextChangedListener(@Nullable OnChangedListener listener);

    /**
     * Returns list of all <b>hashtags</b> found in {@link TextView#getText()}.
     */
    @NonNull
    List<String> getHashtags();

    /**
     * Returns list of all <b>mentions</b> found in {@link TextView#getText()}.
     */
    @NonNull
    List<String> getMentions();

    /**
     * Returns list of all <b>hyperlinks</b> found in {@link TextView#getText()}.
     */
    @NonNull
    List<String> getHyperlinks();

    /**
     * Interface definition for a callback to be invoked when a <b>hashtag</b>,
     * <b>mention</b>, or <b>hyperlink</b> is clicked.
     */
    interface OnClickListener {
        /**
         * Called when a text has been clicked.
         *
         * @param view The view that the texts belong to.
         * @param text The text that was clicked.
         */
        void onClick(@NonNull SocialView view, @NonNull CharSequence text);
    }

    /**
     * Interface definition for a callback to be invoked when a <b>hashtag</b>,
     * <b>mention</b>, or <b>hyperlink</b> is modified.
     */
    interface OnChangedListener {
        /**
         * Called when a text has been modified.
         *
         * @param view The view that the texts belong to.
         * @param text The text that was modified.
         */
        void onChanged(@NonNull SocialView view, @NonNull CharSequence text);
    }
}

File: socialview/src/com/hendraanggrian/appcompat/widget/SocialView.kt
package com.hendraanggrian.appcompat.widget

import android.content.res.ColorStateList
import android.content.res.ColorStateList.valueOf
import android.util.AttributeSet
import android.widget.TextView
import androidx.core.util.PatternsCompat.WEB_URL
import kotlin.text.RegexOption.IGNORE_CASE

/**
 * Base methods of all socialview's widgets.
 * The implementation, however, are kept in [SocialViewImpl].
 *
 * @see SocialViewImpl
 */
interface SocialView<T : TextView> {

    var flags: Int

    /**
     * Internal function to initialize [TextView] with [SocialView].
     * To be called immediately upon view creation.
     *
     * @param attrs from view's constructor.
     */
    fun T.setup(attrs: AttributeSet?)

    /** Determine whether this view should find and span hashtags. */
    fun isHashtagEnabled(): Boolean = FLAG_HASHTAG.isPresent()

    /** Enable or disable hashtag spanning, if not already. */
    fun setHashtagEnabled(enabled: Boolean) = FLAG_HASHTAG.modify(enabled, isHashtagEnabled())

    /** Determine whether this view should find and span mentions. */
    fun isMentionEnabled(): Boolean = FLAG_MENTION.isPresent()

    /** Enable or disable mention spanning, if not already. */
    fun setMentionEnabled(enabled: Boolean) = FLAG_MENTION.modify(enabled, isMentionEnabled())

    /** Determine whether this view should find and span hyperlinks. */
    fun isHyperlinkEnabled(): Boolean = FLAG_HYPERLINK.isPresent()

    /** Enable or disable hyperlink spanning, if not already. */
    fun setHyperlinkEnabled(enabled: Boolean) = FLAG_HYPERLINK.modify(enabled, isHyperlinkEnabled())

    /** Color of hashtag spans. Default is color accent of current app theme. */
    var hashtagColors: ColorStateList

    /** Color of mention spans. Default is color accent of current app theme. */
    var mentionColorStateList: ColorStateList

    /** Color of hyperlink spans. Default is color accent of current app theme. */
    var hyperlinkColors: ColorStateList

    /** Get and set hashtag color from color integer. */
    var hashtagColor: Int
        get() = hashtagColors.defaultColor
        set(value) {
            hashtagColors = valueOf(value)
        }

    /** Get and set mention color from color integer. */
    var mentionColor: Int
        get() = mentionColorStateList.defaultColor
        set(value) {
            mentionColorStateList = valueOf(value)
        }

    /** Get and set hyperlink color from color integer. */
    var hyperlinkColor: Int
        get() = hyperlinkColors.defaultColor
        set(value) {
            hyperlinkColors = valueOf(value)
        }

    /** Register a callback to be invoked when hashtag is clicked. */
    fun setOnHashtagClickListener(listener: ((view: T, String) -> Unit)?)

    /** Register a callback to be invoked when mention is clicked. */
    fun setOnMentionClickListener(listener: ((view: T, String) -> Unit)?)

    /** Register a callback to be invoked when hyperlink is clicked. */
    fun setOnHyperlinkClickListener(listener: ((view: T, String) -> Unit)?)

    /** Register a text watcher to be invoked when hashtag is modified. */
    fun setHashtagTextChangedListener(watcher: ((view: T, String) -> Unit)?)

    /** Register a text watcher to be invoked when mention is modified. */
    fun setMentionTextChangedListener(watcher: ((view: T, String) -> Unit)?)

    /** Obtain all hashtags in current text. */
    val hashtags: List<String>

    /** Obtain all mentions in current text. */
    val mentions: List<String>

    /** Obtain all hyperlinks in current text. */
    val hyperlinks: List<String>

    /** Internal method to span the text based on current configuration. */
    fun colorize()

    @Suppress("NOTHING_TO_INLINE")
    private inline fun Int.isPresent() = flags or this == flags

    @Suppress("NOTHING_TO_INLINE")
    private fun Int.modify(enabled: Boolean, initialState: Boolean) {
        if (enabled != initialState) {
            flags = when {
                enabled -> flags or this
                else -> flags and inv()
            }
            colorize()
        }
    }

    companion object {
        internal const val FLAG_HASHTAG = 1
        internal const val FLAG_MENTION = 2
        internal const val FLAG_HYPERLINK = 4

        internal var REGEX_HASHTAG: Regex = "#(\\w+)".toRegex()
        internal var REGEX_MENTION: Regex = "@(\\w+)".toRegex()
        internal var REGEX_HYPERLINK: Regex = WEB_URL.toRegex()

        /** Change current hashtag regex. */
        fun setHashtagRegex(regex: String) {
            REGEX_HASHTAG = regex.toRegex(IGNORE_CASE)
        }

        /** Change current mention regex. */
        fun setMentionRegex(regex: String) {
            REGEX_MENTION = regex.toRegex(IGNORE_CASE)
        }

        /** Change current hyperlink regex. */
        fun setHyperlinkRegex(regex: String) {
            REGEX_HYPERLINK = regex.toRegex(IGNORE_CASE)
        }
    }
}